taipy/core/data/data_node.py

```
@@ -373,6 +373,28 @@ def _get_last_modified_datetime(cls, path: Optional[str] = None) -> Optional[dat
     def storage_type(cls) -> str:
         raise NotImplementedError
 
+    def _get_downloadable_path(self) -> str:
```

this has already defined in the FileDataNodeMixin no?

---

Yes. The idea is that we want the method to only defined in `_ FileDataNodeMixin`.

`SQLDataNode. _get_downloadable_path()` should raise the NotImplementedError error.

---

ah ok, then I would suggest to move this code in to _abstract_sql.py instead, since DataNode and _FileDataNodeMixin are 2 different abstract class, I don't think they will have much connection in this sense.

---

No I think I misrepresented my idea.

So, for all data nodes that are not File-based, calling to `_get_downloadable_path` should raise an error.
For file-based dn, it should work.

---

hmm in that case, I think it should still be fine if you put it in the _abstract_sql, since other DataNode are not extending from `_FileDataNodeMixin` so it should be fine for them no?

---

what about other classes like Mongo, S3, Generic, ...?
If we do that, we also need to add one for each of them, which is weird.

---

hmmm I don't think they are extending the class `_FileDataNodeMixin` :/